Implement taking backup in --test_mode

As described in #99 and #101 we have implemented the server setup/benchmark run things.
And now it is turn of taking backup.

We have this module, but it seems to be it needs some modifications.

```
=============================================================================== FAILURES ================================================================================
__________________________________________________________________ TestTakeBackup.test_run_all_backup ___________________________________________________________________

self = <test.test_take_backup.TestTakeBackup object at 0x7f81e4d0b828>
return_take_backup_obj = <prepare_env_test_mode.take_backup.WrapperForBackupTest object at 0x7f81e4dac898>

    def test_run_all_backup(self, return_take_backup_obj):
>       return_take_backup_obj.run_all_backup()

test/test_take_backup.py:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
prepare_env_test_mode/take_backup.py:16: in run_all_backup
    self.backup_obj.all_backup()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <master_backup_script.backuper.Backup object at 0x7f81e587d9e8>

    def all_backup(self):
        """
             This function at first checks full backup directory, if it is empty takes full backup.
             If it is not empty then checks for full backup time.
             If the recent full backup  is taken 1 day ago, it takes full backup.
             In any other conditions it takes incremental backup.
            """
        # Workaround for circular import dependency error in Python
    
        # Creating object from CheckEnv class
        check_env_obj = CheckEnv(self.conf)
    
        if check_env_obj.check_all_env():
    
            if self.recent_full_backup_file() == 0:
                logger.debug(
                    "###############################################################")
                logger.debug(
                    "#You have no backups : Taking very first Full Backup! - - - - #")
                logger.debug(
                    "###############################################################")
    
                time.sleep(3)
    
                # Flushing Logs
                if self.mysql_connection_flush_logs():
    
                    # Taking fullbackup
                    if self.full_backup():
                        # Removing old inc backups
                        self.clean_inc_backup_dir()
    
                # Copying backups to remote server
                if hasattr(
                        self,
                        'remote_conn') and hasattr(
                        self,
                        'remote_dir') and self.remote_conn and self.remote_dir:
                    self.copy_backup_to_remote_host()
    
                # Exiting after taking full backup
>               exit(0)
E               SystemExit: 0

master_backup_script/backuper.py:680: SystemExit

```

It seems to be wrong config file options were generated

--- a/general_conf/bck.conf
+++ b/general_conf/bck.conf
@@ -85,3 +85,4 @@ chown_command=chown -R mysql:mysql
 [TestConf]
 gitcmd=--recursive --depth=1 https://github.com/percona/percona-server.git -b 5.7
 testpath=/home/shahriyar.rzaev/XB_TEST/server_dir
+incremental_count=3

--- a/general_conf/check_env.py
+++ b/general_conf/check_env.py
@@ -46,7 +46,7 @@ class CheckEnv(GeneralClass):
             return True
 
     def check_mysql_conf(self):
-        if not os.path.exists(self.mycnf):
+        if not os.path.exists(self.mycnf) and (self.mycnf is not None):
             # Testing with MariaDB Galera Cluster
             # if not os.path.exists(self.maria_cluster_cnf):
             logger.error(

--- a/general_conf/check_env.py
+++ b/general_conf/check_env.py
@@ -46,7 +46,10 @@ class CheckEnv(GeneralClass):
             return True
 
     def check_mysql_conf(self):
-        if not os.path.exists(self.mycnf) and (self.mycnf is not None):
+        if self.mycnf is None:
+            logger.debug("Skipping my.cnf check, because it is not specified")
+            return True
+        elif not os.path.exists(self.mycnf) and (self.mycnf is not None):
             # Testing with MariaDB Galera Cluster
             # if not os.path.exists(self.maria_cluster_cnf):
             logger.error(

--- a/general_conf/check_env.py
+++ b/general_conf/check_env.py
@@ -46,7 +46,7 @@ class CheckEnv(GeneralClass):
             return True
 
     def check_mysql_conf(self):
-        if self.mycnf is None:
+        if self.mycnf is None or self.mycnf == '':
             logger.debug("Skipping my.cnf check, because it is not specified")
             return True
         elif not os.path.exists(self.mycnf) and (self.mycnf is not None):

--- a/general_conf/generalops.py
+++ b/general_conf/generalops.py
@@ -126,6 +126,7 @@ class GeneralClass:
             TEST = con['TestConf']
             self.gitcmd = TEST['gitcmd']
             self.testpath = TEST['testpath']
+            self.incremental_count = TEST['incremental_count']
 
         else:
             logger.critical("Missing config file : /etc/bck.conf")

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -33,7 +33,8 @@ class Backup(GeneralClass):
         self.conf = config
         self.dry = dry_run
         # Call GeneralClass for storing configuration
-        GeneralClass.__init__(self, self.conf)
+        super().__init__(self.conf)
+        #GeneralClass.__init__(self, self.conf)
 
     def sorted_ls(self, path):
         mtime = lambda f: os.stat(os.path.join(path, f)).st_mtime

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -719,7 +719,7 @@ class Backup(GeneralClass):
                     self.copy_backup_to_remote_host()
 
                 # Exiting after taking NEW full backup
-                exit(0)
+                #exit(0)
 
             else:
                 logger.debug(
@@ -746,7 +746,7 @@ class Backup(GeneralClass):
                     self.copy_backup_to_remote_host()
 
                 # Exiting after taking Incremental backup
-                exit(0)
+                #exit(0)
 
 # b = Backup()
 # b.all_backup()

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -677,7 +677,7 @@ class Backup(GeneralClass):
                     self.copy_backup_to_remote_host()
 
                 # Exiting after taking full backup
-                exit(0)
+                #exit(0)
 
             elif self.last_full_backup_date() == 1:
                 logger.debug(

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -92,9 +92,12 @@ class Backup(GeneralClass):
         It is highly recomended to flush binary logs before each full backup for easy maintenance.
         That's why we will execute "flush logs" command before each full backup!
         """
-
-        command_connection = '{} --defaults-file={} -u{} --password={} --host={}'
-        command_execute = ' -e "flush logs"'
+        if hasattr(self, 'mysql_socket'):
+            command_connection = '{} --defaults-file={} -u{} --password={}'
+            command_execute = ' -e "flush logs"'
+        else:
+            command_connection = '{} --defaults-file={} -u{} --password={} --host={}'
+            command_execute = ' -e "flush logs"'
 
         # Open connection
 

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -107,7 +107,7 @@ class Backup(GeneralClass):
                 self.mycnf,
                 self.mysql_user,
                 self.mysql_password,
-                self.mysql_host,
+                #self.mysql_host,
                 self.mysql_socket
             )
         else:

--- a/prepare_env_test_mode/config_generator.py
+++ b/prepare_env_test_mode/config_generator.py
@@ -40,7 +40,7 @@ class ConfigGenerator(CloneBuildStartServer):
                     config.add_section(section2)
                     config.set(section2, "#Optional: set pid directory")
                     config.set(section2, "pid_dir", "/tmp/MySQL-AutoXtraBackup")
-                    config.set(section2, "tmp_dir", "/home/shahriyar.rzaev/XB_TEST/mysql_datadirs")
+                    config.set(section2, "tmpdir", "/home/shahriyar.rzaev/XB_TEST/mysql_datadirs")
                     config.set(section2, "#Optional: set warning if pid of backup us running for longer than X")
                     config.set(section2, "pid_runtime_warning", "2 Hours")
                     config.set(section2, "backupdir", "/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_7") # Can be changed in the future depending on PS version

--- a/prepare_env_test_mode/config_generator.py
+++ b/prepare_env_test_mode/config_generator.py
@@ -102,8 +102,8 @@ class ConfigGenerator(CloneBuildStartServer):
                     section6 = "Remote"
                     config.add_section(section6)
                     config.set(section6, "#Optional remote syncing")
-                    config.set(section6, "remote_conn", "root@xxx.xxx.xxx.xxx")
-                    config.set(section6, "remote_dir", "/home/sh/Documents")
+                    config.set(section6, "#remote_conn", "root@xxx.xxx.xxx.xxx")
+                    config.set(section6, "#remote_dir", "/home/sh/Documents")
 
                     section7 = "Commands"
                     config.add_section(section7)

--- a/prepare_env_test_mode/config_generator.py
+++ b/prepare_env_test_mode/config_generator.py
@@ -115,6 +115,7 @@ class ConfigGenerator(CloneBuildStartServer):
                     config.add_section(section8)
                     config.set(section8, "gitcmd", "--recursive --depth=1 https://github.com/percona/percona-server.git -b 5.7")
                     config.set(section8, "testpath", "/home/shahriyar.rzaev/XB_TEST/server_dir")
+                    config.set(section8, "incremental_count", "3")
 
                     config.write(cfgfile)
 

--- a/prepare_env_test_mode/runner_test_mode.py
+++ b/prepare_env_test_mode/runner_test_mode.py
@@ -1,4 +1,5 @@
 from prepare_env_test_mode.clone_build_start_server import CloneBuildStartServer
+from prepare_env_test_mode.config_generator import ConfigGenerator
 from prepare_env_test_mode.run_benchmark import RunBenchmark
 import os
 import logging
@@ -12,17 +13,30 @@ class RunnerTestMode:
 
     def all_runner(self):
         if self.basedir and (os.path.isfile("{}/all_no_cl".format(self.basedir))):
+            logger.debug("It seems to be the test setup already done...")
             if self.clone_obj.wipe_server_all(self.basedir):
                 if RunBenchmark().run_sysbench():
                     # Take backup/prepare and recover here
                     pass
 
         else:
+            logger.debug("Starting test setup from scratch...")
             if self.clone_obj.clone_percona_qa():
                 if self.clone_obj.clone_ps_server_from_conf():
                     if self.clone_obj.build_server():
                         base_dir = self.clone_obj.get_basedir()
                         if self.clone_obj.prepare_startup(basedir_path=base_dir):
                             if self.clone_obj.start_server(basedir_path=base_dir):
-                                # Take backup/prepare and recover here
-                                pass
+                                url_2_4 = "http://jenkins.percona.com/view/QA/job/qa.pxb24.build/BUILD_TYPE=debug,label_exp=centos7-64/lastSuccessfulBuild/artifact/target/percona-xtrabackup-2.4.x-debug.tar.gz"
+                                url_2_3 = "http://jenkins.percona.com/view/QA/job/qa.pxb23.build/BUILD_TYPE=debug,label_exp=centos7-64/lastSuccessfulBuild/artifact/target/percona-xtrabackup-2.3.x-debug.tar.gz"
+                                if self.clone_obj.get_xb_packages(url_2_4[-37:], url_2_4):
+                                    if self.clone_obj.get_xb_packages(url_2_3[-37:], url_2_3):
+                                        archive_2_4 = "percona-xtrabackup-2.4.x-debug.tar.gz"
+                                        archive_2_3 = "percona-xtrabackup-2.3.x-debug.tar.gz"
+                                        if self.clone_obj.extract_xb_archive(file_name=archive_2_4):
+                                            if self.clone_obj.extract_xb_archive(file_name=archive_2_3):
+                                                conf_obj = ConfigGenerator()
+                                                if conf_obj.generate_config_files():
+                                                    if RunBenchmark().run_sysbench():
+                                                        # Take backup/prepare and recover here
+                                                        pass

--- a/prepare_env_test_mode/runner_test_mode.py
+++ b/prepare_env_test_mode/runner_test_mode.py
@@ -1,6 +1,7 @@
 from prepare_env_test_mode.clone_build_start_server import CloneBuildStartServer
 from prepare_env_test_mode.config_generator import ConfigGenerator
 from prepare_env_test_mode.run_benchmark import RunBenchmark
+from prepare_env_test_mode.take_backup import WrapperForBackupTest
 import os
 import logging
 logger = logging.getLogger(__name__)
@@ -16,7 +17,8 @@ class RunnerTestMode:
             logger.debug("It seems to be the test setup already done...")
             if self.clone_obj.wipe_server_all(self.basedir):
                 if RunBenchmark().run_sysbench():
-                    # Take backup/prepare and recover here
+                    # Take backup
+                    WrapperForBackupTest().run_all_backup()
                     pass
 
         else:
@@ -38,5 +40,6 @@ class RunnerTestMode:
                                                 conf_obj = ConfigGenerator()
                                                 if conf_obj.generate_config_files():
                                                     if RunBenchmark().run_sysbench():
-                                                        # Take backup/prepare and recover here
+                                                        # Take backup
+                                                        WrapperForBackupTest().run_all_backup()
                                                         pass

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -7,7 +7,7 @@ class WrapperForBackupTest(CloneBuildStartServer):
 
     def __init__(self):
         super().__init__()
-        self.backup_obj = Backup(config="{}/xb_2_4.conf")
+        self.backup_obj = Backup(config="{}/xb_2_4.conf".format(self.testpath)) 
 
 
     def run_all_backup(self):

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -1,5 +1,4 @@
 from master_backup_script.backuper import Backup
-from general_conf.generalops import GeneralClass
 from prepare_env_test_mode.clone_build_start_server import CloneBuildStartServer
 
 

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -1,10 +1,10 @@
 from master_backup_script.backuper import Backup
 from general_conf.generalops import GeneralClass
 
+
 class WrapperForBackupTest(Backup):
 
     def __init__(self):
-        self.testpath = GeneralClass().testpath
         super().__init__(config="{}/xb_2_4.conf".format(self.testpath))
 
 

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -7,10 +7,10 @@ class WrapperForBackupTest(CloneBuildStartServer):
 
     def __init__(self):
         super().__init__()
-        self.backup_obj = Backup(config="{}/xb_2_4.conf".format(self.testpath)) 
+        self.backup_obj = Backup(config="{}/xb_2_4.conf".format(self.testpath))
 
 
     def run_all_backup(self):
         # Method for taking backups using backuper.py::all_backup()
-        for _ in range(self.backup_obj.incremental_count + 1):
+        for _ in range(int(self.backup_obj.incremental_count) + 1):
             self.backup_obj.all_backup()

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -1,9 +1,10 @@
 from master_backup_script.backuper import Backup
-
+from general_conf.generalops import GeneralClass
 
 class WrapperForBackupTest(Backup):
 
     def __init__(self):
+        self.testpath = GeneralClass().testpath
         super().__init__(config="{}/xb_2_4.conf".format(self.testpath))
 
 

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -1,14 +1,16 @@
 from master_backup_script.backuper import Backup
 from general_conf.generalops import GeneralClass
+from prepare_env_test_mode.clone_build_start_server import CloneBuildStartServer
 
 
-class WrapperForBackupTest(Backup):
+class WrapperForBackupTest(CloneBuildStartServer):
 
     def __init__(self):
-        super().__init__(config="{}/xb_2_4.conf".format(self.testpath))
+        super().__init__()
+        self.backup_obj = Backup(config="{}/xb_2_4.conf")
 
 
     def run_all_backup(self):
         # Method for taking backups using backuper.py::all_backup()
-        for _ in range(self.incremental_count + 1):
-            self.all_backup()
+        for _ in range(self.backup_obj.incremental_count + 1):
+            self.backup_obj.all_backup()

--- a/prepare_env_test_mode/take_backup.py
+++ b/prepare_env_test_mode/take_backup.py
@@ -1,0 +1,13 @@
+from master_backup_script.backuper import Backup
+
+
+class WrapperForBackupTest(Backup):
+
+    def __init__(self):
+        super().__init__(config="{}/xb_2_4.conf".format(self.testpath))
+
+
+    def run_all_backup(self):
+        # Method for taking backups using backuper.py::all_backup()
+        for _ in range(self.incremental_count + 1):
+            self.all_backup()

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -2,6 +2,7 @@ import pytest
 from prepare_env_test_mode.clone_build_start_server import CloneBuildStartServer
 from prepare_env_test_mode.run_benchmark import RunBenchmark
 from prepare_env_test_mode.config_generator import ConfigGenerator
+from prepare_env_test_mode.take_backup import WrapperForBackupTest
 
 clb_obj = CloneBuildStartServer()
 
@@ -29,3 +30,10 @@ cg_obj = ConfigGenerator()
 @pytest.fixture()
 def return_config_generator_obj():
     return cg_obj
+
+
+bck_obj = WrapperForBackupTest()
+
+@pytest.fixture()
+def return_take_backup_obj():
+    return bck_obj

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -31,9 +31,7 @@ cg_obj = ConfigGenerator()
 def return_config_generator_obj():
     return cg_obj
 
-
-bck_obj = WrapperForBackupTest()
-
 @pytest.fixture()
 def return_take_backup_obj():
+    bck_obj = WrapperForBackupTest()
     return bck_obj

--- a/test/test_take_backup.py
+++ b/test/test_take_backup.py
@@ -1,0 +1,8 @@
+import pytest
+
+@pytest.mark.usefixtures("return_take_backup_obj")
+class TestTakeBackup:
+    # Test class for take_backup.py
+
+    def test_run_all_backup(self, return_take_backup_obj):
+        return_take_backup_obj.run_all_backup()

--- a/test/test_take_backup.py
+++ b/test/test_take_backup.py
@@ -5,7 +5,7 @@ class TestTakeBackup:
     # Test class for take_backup.py
 
     def test_run_all_backup(self, return_take_backup_obj):
-        with pytest.raises(SystemExit) as exit_status:
-            return_take_backup_obj.run_all_backup()
-        assert exit_status.type == SystemExit
-        assert exit_status.value.code == 0
+        #with pytest.raises(SystemExit) as exit_status:
+        return_take_backup_obj.run_all_backup()
+        #assert exit_status.type == SystemExit
+        #assert exit_status.value.code == 0

--- a/test/test_take_backup.py
+++ b/test/test_take_backup.py
@@ -5,4 +5,7 @@ class TestTakeBackup:
     # Test class for take_backup.py
 
     def test_run_all_backup(self, return_take_backup_obj):
-        return_take_backup_obj.run_all_backup()
+        with pytest.raises(SystemExit) as exit_status:
+            return_take_backup_obj.run_all_backup()
+        assert exit_status.type == SystemExit
+        assert exit_status.value.code == 0

